igvm/hypervisor_preferences.py

```diff
@@ -66,209 +96,336 @@ def __repr__(self):
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv) -> bool:
-        # New VM has no hypervisor attribute yet.
+    def get_score(self, vm, hv) -> Union[float, bool]:
+        # New VM has no hypervisor attribute, yet, so we cannot calculate a
+        # score here. We will just allow all HVs to take that VM for now.
         if not vm.hypervisor:
-            return False
+            return True
 
         hv_model = hv.dataset_obj[self.hardware_model]
 
-        # Bail out if hardware_model is not in HYPERVISOR_CPU_THRESHOLDS list
+        # Bail out if hardware_model is not in HYPERVISOR_CPU_THRESHOLDS list.
         if hv_model not in self.hv_cpu_thresholds:
             log.warning(
                 'Missing setting for "{}" in HYPERVISOR_CPU_THRESHOLDS'.format(
-                    hv_model))
+                    hv_model,
+                ),
+            )
             return False
 
         hv_cpu_threshold = float(self.hv_cpu_thresholds[hv_model])
         hv_cpu_util_overall = hv.estimate_cpu_usage(vm)
 
-        return (
-            hv_cpu_util_overall is not None and
-            hv_cpu_util_overall > hv_cpu_threshold
-        )
+        # If there is no value we assume it's a fresh hv.
+        if hv_cpu_util_overall is None:
+            return True
+
+        # Since this is a limiting preference, we don't want any vm end up on
+        # the hv that would exceed the cpu threshold
+        if hv_cpu_util_overall > hv_cpu_threshold:
+            return False
+
+        # Normalize the value. This is only valid for percentage values like
+        # cpu_util_pct and iops_avg. Arbitrary numbers are somewhat difficult
+        # to put into context.
+        return 1 - (hv_cpu_util_overall / hv_cpu_threshold)
 
 
-class HypervisorEnvironmentValue():
+class HypervisorEnvironmentValue(HypervisorPreference):
     """Check if the environment of the hypervisor fits with the VM env.
 
     Make any hypervisor less likely chosen, which would have a different
     environment.
     """
-    def __init__(self, hv_env: str):
-        self.hv_env = hv_env
 
-    def __repr__(self):
+    def __init__(self, hv_env: str) -> None:
+        self.hv_env: str = hv_env
+
+    def __repr__(self) -> str:
         args = repr(self.hv_env)
 
         return '{}({})'.format(type(self).__name__, args)
 
-    def __call__(self, vm, hv) -> bool:
+    def get_score(self, vm, hv) -> Union[float, bool]:
         hypervisor_env = hv.dataset_obj[self.hv_env]
         vm_env = vm.dataset_obj['environment']
 
+        # If the environment is matching, this is our candidate!
         if hypervisor_env == vm_env:
-            return True
+            return 1.
+
+        # Unfortunately we don't always have enough HVs to ensure the same
+        # environment. Therefore we will just highly discourage the hv
+        # instead of excluding it.
+        return 0.01
```

Comment:
I believe that `environment` should be a hard criterion. Not having enough HVs is an issue that needs to be addressed by reassigning or moving hardware around, not by moving VMs.

Reply:
but in af for example we don't have a single staging hv. we have a parameter to enforce this: `--enforce-vm-env`. this should not be a hard criterion imo

Reply:
I don't think people will be very happy if I reboot a HV just because it's not production and there are production VMs on it.

Reply:
i would like to get this version of the pr merged. i have not changed any logic here. it was the same before. let's create a ticket for this and then we can think about how to solve it properly. long story short: out of scope